python_modules/dagster/dagster/core/definitions/graph_definition.py

```diff
@@ -728,7 +772,11 @@ def _validate_in_mappings(
             target_type = target_input.dagster_type
             fan_in_msg = ""
 
-        if target_type != mapping.definition.dagster_type and class_name != "GraphDefinition":
+        if (
+            # no need to check mapping type for graphs because users can't specify ins/out type on graphs
+            class_name not in (GraphDefinition.__name__, SubselectedGraphDefinition.__name__)
+            and target_type != mapping.definition.dagster_type
+        ):
```

Comment:
tangent:

should we invert this condition to only check on composite solid? 

even further if we removed this restriction for graph, why enforce it at all? 

did we add anything that verifies that all the targets of a mapping have the same type ? 

```
@op
def num(x: int):
  ...

@op
def word(y: str):
  ...

@graph
def wat(z):
  # do we catch this early or just let it fail at runtime ?
  num(z)
  word(z)

```

Reply:
good catch! tested it just now and found `wat` would actually pass because x and y are both `Any`. will cut a separate branch to address that.